### DISCUSSION TITLE
Is Agenda not atomic?

### DISCUSSION THREAD

**Discussion by mark-sh-ts** [agenda/agenda#1246] [Q&A]
The comparison table with Bull and Bee says Agenda doesn't have "Atomic ops". The code itself looks pretty atomic though - all locks and updates are always done via findOneAndUpdate. Does "Atomic ops" mean something else in this context? Or is there an atomicity problem i'm not seeing?

**Comment by harisvsulaiman** [comment]
Agenda is not atomic at the moment. Although there is an effort to make it so. #715 #821  (stale).

**Reply by mark-sh-ts** [reply]
Those PRs are about supporting external transactions ( as described [here](https://github.com/agenda/agenda/pull/715#issuecomment-557970455)). I'm talking about internal atomicity. An example of non-atomic code would be:
```javascript
export class Job {
  async complete() {
    this.completed = true
    await mongo.save(this) 
    this.completionTime = new Date() // an error or unrelated app crash here will leave this job with completed=true but without completionDate
    await mongo.save(this)
  }
}
```
I didn't see anything like that in the code, and it looks perfectly atomic, hence my confusion. It seems like the table wrongly describes a limitation that the code doesn't actually have.

**Comment by harisvsulaiman** [comment]
@mark-sh-ts There is an ongoing effort to support transactions in agenda See #1372
